asapdiscovery-ml/asapdiscovery/ml/weights/weights.py

```diff
@@ -0,0 +1,151 @@
+import logging
+from pathlib import Path
+from typing import Dict, List, Union  # noqa: F401
+
+import requests
+import validators
+import yaml
+
+
+def fetch_weights_from_spec(
+    yamlfile: str,
+    models: Union[list[str], str],
+    local_dir: str = "./_weights/",
+    force_fetch: bool = False,
+) -> dict[str, Path]:
+    """Fetch weights from yaml spec file.
+
+    Parameters
+    ----------
+    yamlfile : str
+        Path to yaml spec file.
+    models : List[str]
+        Model names to fetch weights for.
+    local_dir : str, default="./_weights/"
+        Local path to save weights if a remote url is provided. or to check if weights exist locally, by default "./_weights/"
+    force_fetch : bool, default=False
+        Force fetch weights from remote, by default False
+
+    Raises
+    ------
+    FileNotFoundError
+        If YAML spec file does not exist.
+
+    Returns
+    -------
+    Dict of model names and paths to weights files.
+    """
+    if not Path(yamlfile).exists():
+        raise FileNotFoundError(f"Yaml spec file {yamlfile} does not exist")
+
+    with open(yamlfile) as f:
+        spec = yaml.safe_load(f)
+
+    if isinstance(models, str):
+        models = [models]
+
+    weights_files = {}
+    filename_set = set()
+    for model in models:
+        model_spec = spec[model]
+        weights = model_spec["weights"]
+        if weights in filename_set:
+            raise ValueError(
+                f"Duplicate file {weights} in spec file. Please specify a unique filename for each model."
+            )
```

Comment:
sorry, I should have caught this before, but isn't this checking if the URL in `weights` has already been downloaded? if so, I think we should probably change this error to just be a warning or something with `logging`

Reply:
This checks that there are not two of the same key in the `yaml` file, perhaps I should change the name to something less confusing.

Reply:
I have changed it to `weights_set` and left a comment